File: Homework/Hang_Nguyen/lesson_1/range-sum-query-2d-immutable.py
class NumMatrix:

    def __init__(self, matrix):
        rows =  len(matrix)
        cols = len(matrix[0])
        self.dp = [[0] * (cols + 1) for r in range(rows + 1)]
        
        for r in range(0, rows):
            prefix = 0
            for c in range(0, cols):
                prefix += matrix[r][c]
                above = self.dp[r][c+1]
                self.dp[r+1][c+1] = prefix + above
        

    def sumRegion(self, row1, col1, row2, col2):
        r1, r2 = row1 + 1, row2 + 1
        c1, c2 = col1 + 1, col2 + 1

        bottomRightRegion = self.dp[r2][c2]
        aboveRegion = self.dp[r1 - 1][c2]
        leftRegion = self.dp[r2][c1 - 1]
        topLeftRegion = self.dp[r1 - 1][c1 - 1]

        return bottomRightRegion - aboveRegion - leftRegion + topLeftRegion


# TEST CASE
numMatrix = NumMatrix([[3, 0, 1, 4, 2], [5, 6, 3, 2, 1], [1, 2, 0, 1, 5], [4, 1, 0, 1, 7], [1, 0, 3, 0, 5]])

t = numMatrix.sumRegion(2, 1, 4, 3)
print("TEST 1 is " + str(t == 8))

t = numMatrix.sumRegion(1, 1, 2, 2)
print("TEST 2 is " + str(t == 11))

t = numMatrix.sumRegion(1, 2, 2, 4)
print("TEST 3 is " + str(t == 12))

File: YangTuanAnh/lesson_1/Range Sum Query 2D - Immutable.cpp
class NumMatrix {
public:
    int a[205][205];
    NumMatrix(vector<vector<int>>& matrix) {
        int n = matrix.size(), m = matrix[0].size();
        memset(a, 0, sizeof(a));
        for (int i=1; i<=n; i++)
            for (int j=1; j<=m; j++)
                a[i][j] = a[i-1][j]+a[i][j-1]-a[i-1][j-1]+matrix[i-1][j-1];

        // for (int i=0; i<=n; i++) {
        //     for (int j=0; j<=m; j++)
        //         cout << a[i][j] << ' ';
        //     cout << '\n';
        // }
    }
    
    int sumRegion(int row1, int col1, int row2, int col2) {
        return a[row2+1][col2+1]-a[row1][col2+1]-a[row2+1][col1]+a[row1][col1];
    }
};

/**
 * Your NumMatrix object will be instantiated and called as such:
 * NumMatrix* obj = new NumMatrix(matrix);
 * int param_1 = obj->sumRegion(row1,col1,row2,col2);
 */